zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/remote/ExecRemoteInterpreterProcess.java

```diff
@@ -0,0 +1,234 @@
+    interpreterProcessLauncher.waitForReady(getConnectTimeout());
+    if (interpreterProcessLauncher.isLaunchTimeout()) {
+      throw new IOException(
+          String.format("Interpreter Process creation is time out in %d seconds", getConnectTimeout() / 1000) + "\n"
+              + "You can increase timeout threshold via "
+              + "setting zeppelin.interpreter.connect.timeout of this interpreter.\n"
+              + interpreterProcessLauncher.getErrorMessage());
+    }
+
+    if (!interpreterProcessLauncher.isRunning()) {
+      throw new IOException("Fail to launch interpreter process:\n" + interpreterProcessLauncher.getErrorMessage());
+    } else {
+      String launchOutput = interpreterProcessLauncher.getProcessLaunchOutput();
+      Matcher m = YARN_APP_PATTER.matcher(launchOutput);
+      if (m.find()) {
+        String appId = m.group(1);
+        LOGGER.info("Detected yarn app: {}, add it to YarnAppMonitor", appId);
+        YarnAppMonitor.get().addYarnApp(ConverterUtils.toApplicationId(appId), this);
+      }
+    }
+  }
+
+  @Override
+  public void processStarted(int port, String host) {
+    super.processStarted(port, host);
+    // for yarn cluster it may be transitioned from COMPLETED to RUNNING.
+    interpreterProcessLauncher.onProcessRunning();
+  }
+
+  @Override
+  public void stop() {
+    if (isRunning()) {
+      super.stop();
+      // wait for a clean shutdown
+      this.interpreterProcessLauncher.waitForShutdown(RemoteInterpreterServer.DEFAULT_SHUTDOWN_TIMEOUT + 500);
+      // kill process
+      this.interpreterProcessLauncher.stop();
+      this.interpreterProcessLauncher = null;
+      LOGGER.info("Remote exec process of interpreter group: {} is terminated", getInterpreterGroupId());
+    } else {
+      LOGGER.warn("Try to stop a not running interpreter process of interpreter group: {}", getInterpreterGroupId());
+    }
+  }
+
+  @VisibleForTesting
+  public String getInterpreterRunner() {
+    return interpreterRunner;
+  }
+
+  @Override
+  public boolean isRunning() {
+    return interpreterProcessLauncher != null && interpreterProcessLauncher.isRunning();
+  }
+
+  @Override
+  public String getErrorMessage() {
+    return this.interpreterProcessLauncher != null
+        ? this.interpreterProcessLauncher.getErrorMessage()
+        : "";
+  }
+
+  private class InterpreterProcessLauncher extends ProcessLauncher {
+
+    public InterpreterProcessLauncher(CommandLine commandLine, Map<String, String> envs) {
+      super(commandLine, envs);
+    }
+
+    public void waitForShutdown(int timeout) {
+      synchronized (this) {
+        long startTime = System.currentTimeMillis();
+        long timeoutTime = startTime + timeout;
+        while (state == State.RUNNING && !Thread.currentThread().isInterrupted()) {
+          long timetoTimeout = timeoutTime - System.currentTimeMillis();
+          if (timetoTimeout <= 0) {
+            LOGGER.warn("Timeout reached");
```

Comment:
Better to specify what timeout reached. e.g. here  it is `Shutdown timeout reached`

Reply:
Fixed